File: packages/cytometry-core/src/transforms/logicle.ts
import type { Transform } from "./types.ts";

/**
 * Logicle (biexponential) transform — Moore W.A. & Parks D.R., "Update for the
 * logicle data scale including operational code implementations," Cytometry
 * Part A 2012;81A(4):273-277 (doi:10.1002/cyto.a.22030). Original method:
 * Parks, Roederer & Moore, Cytometry A 2006;69A(6):541-551.
 *
 * Reference implementation © Board of Trustees, Leland Stanford Junior
 * University, BSD-style license (package edu.stanford.facs.transform). This is a
 * faithful TypeScript port. The forward map is a biexponential
 *
 *     B(x) = a*e^(b*x) - c*e^(-d*x) + f
 *
 * constructed so that B(x1) = 0 (data value 0 sits at scale x1, giving a
 * linear-like region around zero) and B(1) = T (top of scale). `unscale` is a
 * direct evaluation; `scale` inverts B numerically with a bracketed
 * Newton/bisection (RTSAFE) — robust for the whole domain including negatives.
 *
 * PATENT NOTE: the logicle display is covered by US Patent 6,954,722 (Stanford).
 * It ships under a permissive BSD license, is part of the ISAC Gating-ML 2.0
 * standard, and downstream projects report Stanford does not enforce it for flow
 * cytometry. Get legal sign-off before commercial distribution. See
 * docs/DERISKING.md.
 *
 * VALIDATION NOTE: this port passes round-trip, monotonicity, and analytic
 * anchor tests (B(x1)=0, B(1)=T). Before production it must additionally be
 * checked against flowCore/FlowKit golden values — see
 * test/logicle.golden.test.ts for the (stubbed) harness.
 */

const LN10 = Math.LN10;
const TAYLOR_LENGTH = 16;

/**
 * Solve for the parameter `d` of the biexponential given `b` and the
 * (normalized) linearization width `w`. Finds the root of
 *   g(d) = 2*ln(d) + w*d - 2*ln(b) + w*b
 * on (0, b) via RTSAFE (Numerical Recipes): Newton's method safeguarded by
 * bisection. g is strictly increasing, g(0+) -> -inf, g(b) = 2*w*b > 0.
 */
export function solveD(b: number, w: number): number {
  if (w === 0) return b; // degenerate: logicle collapses to asinh
  const tolerance = 2 * b * Number.EPSILON;
  let dLo = 0;
  let dHi = b;
  const fConst = -2 * Math.log(b) + w * b;
  let d = 0.5 * (dLo + dHi);
  let lastDelta = dHi - dLo;
  let f = 2 * Math.log(d) + w * d + fConst;

  for (let i = 0; i < 40; i++) {
    const df = 2 / d + w;
    let delta: number;
    // Bisect if Newton would leave the bracket or is converging too slowly.
    if (
      ((d - dHi) * df - f) * ((d - dLo) * df - f) >= 0 ||
      Math.abs(1.9 * f) > Math.abs(lastDelta * df)
    ) {
      delta = 0.5 * (dHi - dLo);
      const dNew = dLo + delta;
      if (dNew === dLo) return dNew;
      d = dNew;
    } else {
      delta = f / df;
      const prev = d;
      d -= delta;
      if (d === prev) return d;
    }
    if (Math.abs(delta) < tolerance) return d;
    lastDelta = delta;
    f = 2 * Math.log(d) + w * d + fConst;
    if (f < 0) dLo = d;
    else dHi = d;
  }
  return d;
}

export class LogicleTransform implements Transform {
  readonly kind = "logicle";
  readonly T: number;
  readonly W: number;
  readonly M: number;
  readonly A: number;

  // Biexponential parameters.
  private readonly a: number;
  private readonly b: number;
  private readonly c: number;
  private readonly d: number;
  private readonly f: number;
  // Breakpoints in scale space.
  private readonly w: number;
  private readonly x1: number;
  private readonly xTaylor: number;
  // Taylor coefficients of B around x1 (taylor[k] is the coeff of (x-x1)^(k+1)).
  private readonly taylor: Float64Array;

  /**
   * @param T top of data scale (e.g. 262144 = 2^18 for 18-bit data)
   * @param W width of the linearization region, in decades (e.g. 0.5)
   * @param M total number of decades the scale spans (e.g. 4.5)
   * @param A number of additional negative decades to display (e.g. 0)
   */
  constructor(T = 262144, W = 0.5, M = 4.5, A = 0) {
    if (T <= 0) throw new RangeError("logicle: T must be > 0");
    if (W < 0) throw new RangeError("logicle: W must be >= 0");
    if (M <= 0) throw new RangeError("logicle: M must be > 0");
    if (W > M / 2) throw new RangeError("logicle: W must be <= M/2");
    if (A < -W || A > M - 2 * W) throw new RangeError("logicle: A out of range");
    this.T = T;
    this.W = W;
    this.M = M;
    this.A = A;

    this.w = W / (M + A);
    const x2 = A / (M + A);
    this.x1 = x2 + this.w;
    const x0 = x2 + 2 * this.w;
    this.b = (M + A) * LN10;
    this.d = solveD(this.b, this.w);

    const cA = Math.exp(x0 * (this.b + this.d));
    const mfA = Math.exp(this.b * this.x1) - cA / Math.exp(this.d * this.x1);
    this.a = T / (Math.exp(this.b) - mfA - cA / Math.exp(this.d));
    this.c = cA * this.a;
    this.f = -mfA * this.a;

    // Taylor series of B around x1 avoids catastrophic cancellation near 0.
    this.xTaylor = this.x1 + this.w / 4;
    let posCoef = this.a * Math.exp(this.b * this.x1);
    let negCoef = -this.c * Math.exp(-this.d * this.x1);
    const taylor = new Float64Array(TAYLOR_LENGTH);
    for (let i = 0; i < TAYLOR_LENGTH; i++) {
      posCoef *= this.b / (i + 1);
      negCoef *= -this.d / (i + 1);
      taylor[i] = posCoef + negCoef;
    }
    this.taylor = taylor;
  }

  private seriesBiexponential(scale: number): number {
    const t = scale - this.x1;
    const taylor = this.taylor;
    let sum = taylor[TAYLOR_LENGTH - 1];
    for (let i = TAYLOR_LENGTH - 2; i >= 0; i--) sum = sum * t + taylor[i];
    return sum * t;
  }

  /** Evaluate B(x) for x >= x1 (no negative-reflection). */
  private evalB(x: number): number {
    if (x < this.xTaylor) return this.seriesBiexponential(x);
    return this.a * Math.exp(this.b * x) + this.f - this.c * Math.exp(-this.d * x);
  }

  /** B'(x) = a*b*e^(b*x) + c*d*e^(-d*x) > 0 — strictly increasing. */
  private evalBprime(x: number): number {
    return (
      this.a * this.b * Math.exp(this.b * x) +
      this.c * this.d * Math.exp(-this.d * x)
    );
  }

  /** logicle scale -> data value */
  unscale(scale: number): number {
    const negative = scale < this.x1;
    const x = negative ? 2 * this.x1 - scale : scale;
    const value = this.evalB(x);
    return negative ? -value : value;
  }

  /** data value -> logicle scale */
  scale(value: number): number {
    if (value === 0) return this.x1;
    const negative = value < 0;
    const v = negative ? -value : value;

    // Bracket the root of B(x) - v on [x1, hi]; B(x1)=0 <= v, extend hi until
    // B(hi) >= v (handles over-range data above T).
    let lo = this.x1;
    let hi = 1;
    while (this.evalB(hi) < v && hi < 64) hi *= 2;

    let x = 0.5 * (lo + hi);
    let dxOld = hi - lo;
    let dx = dxOld;
    let fx = this.evalB(x) - v;
    let dfx = this.evalBprime(x);
    // Tolerance is on the step in scale space (x in ~[0,1]), NOT on the data
    // residual — mixing data-space v in here makes it far too loose at large T.
    const tol = 1e-13;

    for (let i = 0; i < 60; i++) {
      if (
        ((x - hi) * dfx - fx) * ((x - lo) * dfx - fx) > 0 ||
        Math.abs(2 * fx) > Math.abs(dxOld * dfx)
      ) {
        dxOld = dx;
        dx = 0.5 * (hi - lo);
        x = lo + dx;
        if (lo === x) break;
      } else {
        dxOld = dx;
        dx = fx / dfx;
        const prev = x;
        x -= dx;
        if (prev === x) break;
      }
      if (Math.abs(dx) < tol) break;
      fx = this.evalB(x) - v;
      dfx = this.evalBprime(x);
      if (fx < 0) lo = x;
      else hi = x;
    }
    return negative ? 2 * this.x1 - x : x;
  }
}

File: packages/cytometry-wasm/src/logicle.rs
//! Logicle transform — a faithful Rust mirror of
//! @joeee/cytometry-core/src/transforms/logicle.ts (Moore & Parks 2012). Kept
//! algorithmically identical so the TS reference and the WASM kernel agree to
//! ~1e-6; the TS tests and these tests assert the same anchors and round-trips.

use core::f64::consts::LN_10;

const TAYLOR_LENGTH: usize = 16;

/// Solve for biexponential `d` given `b` and normalized width `w` (RTSAFE).
pub fn solve_d(b: f64, w: f64) -> f64 {
    if w == 0.0 {
        return b;
    }
    let tolerance = 2.0 * b * f64::EPSILON;
    let mut d_lo = 0.0;
    let mut d_hi = b;
    let f_const = -2.0 * b.ln() + w * b;
    let mut d = 0.5 * (d_lo + d_hi);
    let mut last_delta = d_hi - d_lo;
    let mut f = 2.0 * d.ln() + w * d + f_const;

    for _ in 0..40 {
        let df = 2.0 / d + w;
        let delta;
        if ((d - d_hi) * df - f) * ((d - d_lo) * df - f) >= 0.0
            || (1.9 * f).abs() > (last_delta * df).abs()
        {
            delta = 0.5 * (d_hi - d_lo);
            let d_new = d_lo + delta;
            if d_new == d_lo {
                return d_new;
            }
            d = d_new;
        } else {
            delta = f / df;
            let prev = d;
            d -= delta;
            if d == prev {
                return d;
            }
        }
        if delta.abs() < tolerance {
            return d;
        }
        last_delta = delta;
        f = 2.0 * d.ln() + w * d + f_const;
        if f < 0.0 {
            d_lo = d;
        } else {
            d_hi = d;
        }
    }
    d
}

pub struct Logicle {
    a: f64,
    b: f64,
    c: f64,
    d: f64,
    f: f64,
    x1: f64,
    x_taylor: f64,
    taylor: [f64; TAYLOR_LENGTH],
}

impl Logicle {
    pub fn new(t: f64, w: f64, m: f64, a: f64) -> Logicle {
        assert!(t > 0.0 && w >= 0.0 && m > 0.0, "logicle: bad T/W/M");
        assert!(w <= m / 2.0, "logicle: W must be <= M/2");
        assert!(a >= -w && a <= m - 2.0 * w, "logicle: A out of range");

        let wn = w / (m + a);
        let x2 = a / (m + a);
        let x1 = x2 + wn;
        let x0 = x2 + 2.0 * wn;
        let b = (m + a) * LN_10;
        let d = solve_d(b, wn);

        let c_a = (x0 * (b + d)).exp();
        let mf_a = (b * x1).exp() - c_a / (d * x1).exp();
        let a_coef = t / (b.exp() - mf_a - c_a / d.exp());
        let c = c_a * a_coef;
        let f = -mf_a * a_coef;

        let x_taylor = x1 + wn / 4.0;
        let mut pos = a_coef * (b * x1).exp();
        let mut neg = -c * (-d * x1).exp();
        let mut taylor = [0.0f64; TAYLOR_LENGTH];
        for i in 0..TAYLOR_LENGTH {
            pos *= b / (i as f64 + 1.0);
            neg *= -d / (i as f64 + 1.0);
            taylor[i] = pos + neg;
        }

        Logicle { a: a_coef, b, c, d, f, x1, x_taylor, taylor }
    }

    fn series(&self, scale: f64) -> f64 {
        let t = scale - self.x1;
        let mut sum = self.taylor[TAYLOR_LENGTH - 1];
        for i in (0..TAYLOR_LENGTH - 1).rev() {
            sum = sum * t + self.taylor[i];
        }
        sum * t
    }

    fn eval_b(&self, x: f64) -> f64 {
        if x < self.x_taylor {
            self.series(x)
        } else {
            self.a * (self.b * x).exp() + self.f - self.c * (-self.d * x).exp()
        }
    }

    fn eval_bprime(&self, x: f64) -> f64 {
        self.a * self.b * (self.b * x).exp() + self.c * self.d * (-self.d * x).exp()
    }

    pub fn unscale(&self, scale: f64) -> f64 {
        let negative = scale < self.x1;
        let x = if negative { 2.0 * self.x1 - scale } else { scale };
        let v = self.eval_b(x);
        if negative {
            -v
        } else {
            v
        }
    }

    pub fn scale(&self, value: f64) -> f64 {
        if value == 0.0 {
            return self.x1;
        }
        let negative = value < 0.0;
        let v = value.abs();

        let mut lo = self.x1;
        let mut hi = 1.0;
        while self.eval_b(hi) < v && hi < 64.0 {
            hi *= 2.0;
        }

        let mut x = 0.5 * (lo + hi);
        let mut dx_old = hi - lo;
        let mut dx = dx_old;
        let mut fx = self.eval_b(x) - v;
        let mut dfx = self.eval_bprime(x);
        let tol = 1e-13;

        for _ in 0..60 {
            if ((x - hi) * dfx - fx) * ((x - lo) * dfx - fx) > 0.0
                || (2.0 * fx).abs() > (dx_old * dfx).abs()
            {
                dx_old = dx;
                dx = 0.5 * (hi - lo);
                x = lo + dx;
                if lo == x {
                    break;
                }
            } else {
                dx_old = dx;
                dx = fx / dfx;
                let prev = x;
                x -= dx;
                if prev == x {
                    break;
                }
            }
            if dx.abs() < tol {
                break;
            }
            fx = self.eval_b(x) - v;
            dfx = self.eval_bprime(x);
            if fx < 0.0 {
                lo = x;
            } else {
                hi = x;
            }
        }
        if negative {
            2.0 * self.x1 - x
        } else {
            x
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn anchors_and_roundtrip() {
        let lg = Logicle::new(262144.0, 0.5, 4.5, 0.0);
        // Same analytic anchors the TS suite asserts.
        assert!((lg.scale(262144.0) - 1.0).abs() < 1e-6);
        assert!(lg.unscale(lg.scale(0.0)).abs() < 1e-6);
        for v in [-5000.0, -10.0, 0.0, 1.0, 100.0, 10000.0, 262144.0] {
            let back = lg.unscale(lg.scale(v));
            let tol = (v.abs() * 1e-5).max(1e-4);
            assert!((back - v).abs() < tol, "roundtrip failed at {v}: {back}");
        }
    }

    #[test]
    fn monotonic() {
        let lg = Logicle::new(262144.0, 0.5, 4.5, 0.0);
        let mut prev = f64::NEG_INFINITY;
        let mut v = -10000.0;
        while v <= 262144.0 {
            let s = lg.scale(v);
            assert!(s > prev, "not monotonic at {v}");
            prev = s;
            v += 997.0;
        }
    }
}
